src/cargo/util/network.rs

```diff
@@ -0,0 +1,45 @@
+use util::{CargoResult, human, Config};
+use term::color::YELLOW;
+
+/// Wrapper method for network call retry logic.
+///
+/// Retry counts provided by Config object 'net.retry'. Config shell outputs
+/// a warning on per retry.
+///
+/// Closure must return a CargoResult.
+///
+/// Example:
+/// use util::network;
+/// cargo_result = network.with_retry(&config, || something.download());
+pub fn with_retry<T, F>(config: &Config, mut callback: F) -> CargoResult<T>
```

Comment:
I think the signature you want for this method is something along the lines of:

```rust
pub fn with_retry<T, E, F>(config: &Config, mut callback: F) -> CargoResult<T>
    where F: FnMut() -> Result<T, E>,
          E: NetworkError + CargoError
```

This would then `match` on the result of `callback` and call `maybe_spurious` to see whether it is indeed a spurious error or not.